src/services/id-service.ts

```diff
@@ -0,0 +1,102 @@
+import {NextFunction, Request, response, Response} from 'express';
+import {InvalidEntityError, InvalidRequestError} from '../middlewares/error-handler.js';
+import { closeDBConnection, getTableDefinition, prepareDataSource } from '../middlewares/datasource.js';
+import { Mutex } from 'async-mutex';
+import * as config from '../config.js';
+import {date, RecordType} from "zod";
+import {getRecord} from "../config-validator.js";
+
+const mutex = new Mutex();
+
+// create new id
+async function createId(searchCriteria: {}, entityType: string, next: NextFunction, requestId: number) {
+	console.log('******** CREATE CALLED :' + requestId);
+	const repo = await prepareDataSource(getTableDefinition(entityType), requestId, config.dbSync);
+
+	const savedEntity = await repo.save(searchCriteria);
+	await closeDBConnection(next, requestId).then(() => console.log('DB connection closed'));
+	return savedEntity;
+}
+
+// check if an argo id exists for the submitter and program
+async function findId(searchCriteria: {}, entityType: string, next: NextFunction, requestId: number) {
+	console.log('******** FIND CALLED :' + requestId);
+	const schemaInfo = getTableDefinition(entityType);
+	const repo = await prepareDataSource(schemaInfo, requestId, config.dbSync);
+
+	const keys = Object.keys(searchCriteria) as (keyof typeof searchCriteria)[];
+
+	let query = repo
+		.createQueryBuilder(schemaInfo.tablename)
+		.where(schemaInfo.tablename + '.' + keys[0] + '= :' + keys[0], { [keys[0]]: searchCriteria[keys[0]] });
+
+	if (keys.length > 1) {
+		for (let i = 1; i <= keys.length - 1; i++) {
+			query = query.andWhere(schemaInfo.tablename + '.' + keys[i] + '= :' + keys[i], {
+				[keys[i]]: searchCriteria[keys[i]],
+			});
+		}
+	}
+
+	const entity = await query.addSelect([schemaInfo.tablename + '.entityId']).getOne();
+	await closeDBConnection(next, requestId).then(() => console.log('DB connection closed'));
+	return entity;
+}
+
+export async function getId(request: Request, response: Response, next: NextFunction, requestId: number) {
+	const release = await mutex.acquire();
+	try {
+		const entityType = request.params.entityType;
+		validateEntityType(entityType, next);
+		const keyCriteria = getSearchCriteria(entityType, request);
+		validateSearchParams(keyCriteria, next);
+		let id = await findId(keyCriteria, entityType, next, requestId);
+
+		if (!id) {
+			id = await createId(keyCriteria, entityType, next, requestId);
+		}
+		return id;
+	} finally {
+		release();
+	}
+
+}
+
+export async function findIdFor(request: Request, response: Response, next: NextFunction, requestId: number) {
```

Comment:
Same idea as with `getId`, we should not be sending the express objects out of the original route handler. Service layer should be independent of them.